llvm/test/Transforms/InstCombine/fcmp-fadd-select.ll

```diff
@@ -0,0 +1,634 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+; fcmp OGT + fadd + sel => fcmp OGT + sel => fmaxnum
+
+define float @test_fcmp_ogt_fadd_select_constant(float %in) {
```

Comment:
This case fails alive: https://alive2.llvm.org/ce/z/_h-8Lw

You can't preserve the nnan unless it was also present on the fadd and fcmp

Reply:
This change does the below transformation, which seems to be safe https://alive2.llvm.org/ce/z/3XQb24.
```
define float @src_test_fcmp_ogt_fadd_select_constant(float %in) {
   %cmp1 = fcmp ogt float %in, 0.000000e+00
   %add = fadd float %in, 1.000000e+00
   %sel = select nnan nsz i1 %cmp1, float %add, float 1.000000e+00
   ret float %sel
 }
```
**=>**
```
define float @tgt_test_fcmp_ogt_fadd_select_constant(float %in) {
   %cmp1 = fcmp ogt float %in, 0.000000e+00
   %sel = select nnan nsz i1 %cmp1, float %in, float 0.000000e+00
   %add = fadd float %sel, 1.000000e+00
   ret float %add
 }
```

The transformation from `tgt_test_fcmp_ogt_fadd_select_constant` to `llvm.maxnum.f32` is an existing one, as shown here https://godbolt.org/z/6n4Tfrx9e. If this transformation is incorrect, we should probably address it in a separate PR.

Reply:
Yes, this needs to be fixed. The flag handling for just this can be more aggressive

Reply:
@arsenm I've been working on making flags more aggressive for the above transformation. I've a question regarding the arguments of Select and FCmp instruction. Since the Select instruction already has the `nnan` flag, wouldn't that imply that the arguments of FCmp also have this property?

Reply:
Sorry, I'm not sure if I understand this fully. Isn't for this transformation the arguments of `fcmp` and `select` instruction should be the same? If possible could you please provide an example where this is a problem?